File: item-list/src/main/java/com/gurunars/item_list/ChangeMove.java
package com.gurunars.item_list;

import android.support.v7.widget.RecyclerView;

import java.util.List;

import javax.inject.Inject;

class ChangeMove<ItemType extends Item> extends ChangeOfPart<ItemType> {

    @Inject
    private ScrollPositionFetcher scrollPositionFetcher = new ScrollPositionFetcher();

    ChangeMove(ItemType item, int sourcePosition, int targetPosition) {
        super(item, sourcePosition, targetPosition);
    }

    @Override
    public int apply(RecyclerView.Adapter adapter, Scroller scroller, List<ItemType> items,
                     int currentPosition) {
        items.remove(sourcePosition);
        items.add(targetPosition, item);
        adapter.notifyItemMoved(sourcePosition, targetPosition);
        return scrollPositionFetcher.getScrollPosition(sourcePosition, scroller, items);
    }
}

File: item-list/src/main/java/com/gurunars/item_list/ChangeMove.kt
package com.gurunars.item_list

import android.support.v7.widget.RecyclerView

import javax.inject.Inject

internal class ChangeMove<ItemType : Item>(item: ItemType, sourcePosition: Int, targetPosition: Int) : ChangeOfPart<ItemType>(item, sourcePosition, targetPosition) {

    @Inject
    private val scrollPositionFetcher = ScrollPositionFetcher()

    override fun apply(adapter: RecyclerView.Adapter<*>, scroller: Scroller, items: MutableList<ItemType>,
                       currentPosition: Int): Int {
        items.removeAt(sourcePosition)
        items.add(targetPosition, item)
        adapter.notifyItemMoved(sourcePosition, targetPosition)
        return scrollPositionFetcher.getScrollPosition(sourcePosition, scroller, items)
    }
}
